src/Kiota.Builder/CodeDOM/CodeClass.cs

```diff
@@ -63,18 +66,107 @@ public bool IsErrorDefinition
         }
         get => InnerChildElements.Values.OfType<CodeIndexer>().FirstOrDefault();
     }
+    public Boolean IsPropertiesBuilt()
+    {
+        return Kind != CodeClassKind.Model || propertiesBuilt.CurrentCount == 0;
+    }
+    public void WaitForPropertiesBuilt()
+    {
+        if (Kind == CodeClassKind.Model && !Monitor.IsEntered(propertiesBuilt))
+        {
+            propertiesBuilt.Wait();
+        }
+    }
+    public void StartBuildingProperties()
+    {
+        if (Kind == CodeClassKind.Model)
+            Monitor.Enter(propertiesBuilt);
+    }
+    public void PropertiesBuildingDone()
+    {
+        if (Kind == CodeClassKind.Model)
+        {
+            if (!IsPropertiesBuilt())
+                propertiesBuilt.Signal();
```

Comment:
Should we consider checking the return value of `Signal()` and throwing if it doesn't return true. In the event of a future refactor of the class, it would be act as a validation to therefore not fail silently.

Reply:
Good idea. I'll work on that.